test-sbat.c

```diff
@@ -0,0 +1,186 @@
+	assert(strncmp(sbat.entries[1]->component_name, "test2", strlen("test2")) == 0);
+	assert(strncmp(sbat.entries[1]->component_generation, "2", strlen("2")) == 0);
+	assert(strncmp(sbat.entries[1]->vendor_name, "SBAT test2", strlen("SBAT test2")) == 0);
+	assert(strncmp(sbat.entries[1]->vendor_package_name, "acme2", strlen("acme2")) == 0);
+	assert(strncmp(sbat.entries[1]->vendor_version, "2", strlen("2")) == 0);
+	assert(strncmp(sbat.entries[1]->vendor_url, "testURL2", strlen("testURL2")) == 0);
+}
+
+void
+test_parse_sbat(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1,testURL";
+	size_t sbat_size = sizeof(sbat_base);
+	char * buffer = NULL;
+	struct sbat sbat = { 0, NULL };
+
+	assert(parse_sbat(NULL, sbat_size, buffer, &sbat) == EFI_INVALID_PARAMETER);
+	assert(parse_sbat(sbat_base, sbat_size, buffer, NULL) == EFI_INVALID_PARAMETER);
+
+	test_parse_sbat_invalid_sbat_data();
+	test_parse_sbat_valid_sbat_data();
+}
+
+static struct sbat_var *
+create_mock_sbat_var_entry (char* comp_gen, char* comp_name_size, char* comp_name)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+	if (!new_entry)
+		return NULL;
+	INIT_LIST_HEAD(&new_entry->list);
+	CHAR8 *alloc_comp_gen = AllocatePool((strlen(comp_gen)+1)*sizeof(*alloc_comp_gen));
+	if (!alloc_comp_gen)
+		return NULL;
+	CHAR8 *alloc_comp_name_size = AllocatePool((strlen(comp_name_size)+1)*sizeof(*alloc_comp_name_size));
+	if (!alloc_comp_name_size)
+		return NULL;
+	CHAR8 *alloc_comp_name = AllocatePool((strlen(comp_name)+1)*sizeof(*alloc_comp_name));
+	if (!alloc_comp_name)
+		return NULL;
+	CopyMem(alloc_comp_gen, comp_gen, strlen(comp_gen) + 1);
+	CopyMem(alloc_comp_name_size, comp_name_size, strlen(comp_name_size)+1);
+	CopyMem(alloc_comp_name, comp_name, strlen(comp_name)+1);
+	new_entry->component_generation = alloc_comp_gen;
+	new_entry->component_name_size = alloc_comp_name_size;
+	new_entry->component_name = alloc_comp_name;
+	return new_entry;
+}
+
+// TEST 1: .sbat entry in EFI variable
+// Expected Result: Pass
+int
+test_verify_sbat_valid_sbat_pass(void)
+{
+	char sbat_base[] = "test1,1,SBAT test1,acme,1,testURL\n" \
+			   "test2,2,SBAT test2,acme2,2,testURL2\n";
+	size_t sbat_size = sizeof(sbat_base);
+	char *buffer = NULL;
+	struct sbat sbat = { 0 };
+	assert(parse_sbat(sbat_base, sbat_size, buffer, &sbat) == EFI_SUCCESS);
+
+	list_t sbat_entries;
+	INIT_LIST_HEAD(&sbat_entries);
+	struct sbat_var *test_entry;
+	char comp_name[] = "test1";
+	char comp_gen[] = "1";
+	char comp_name_size[5];
+	sprintf(comp_name_size, "%d", strlen(comp_name)+1);
+	test_entry = create_mock_sbat_var_entry(comp_gen, comp_name_size, comp_name);
+	if (!test_entry)
+		return -1;
+	list_add(&test_entry->list, &sbat_entries);
+
+	// Note: verify_sbat also frees the underlying sbat_entries.
+	assert(verify_sbat(&sbat, &sbat_entries) == EFI_SUCCESS);
+	return 0;
+}
+
+// Test: .sbat component generation higher than SBAT EFI generation
```

Comment:
comment is backwards...
Test scenario: .sbat component generation read from binary is lower than the SBAT EFI component generation for this component, therefore verify_sbat() should reject with EFI_SECURITY_VIOLATION.